redash/models.py

```diff
@@ -1000,6 +1006,32 @@ def outdated_queries(cls):
 
         return outdated_queries.values()
 
+    @classmethod
+    def delete_stale_resultsets(cls):
+        delete_count = 0
+        queries = Query.query.filter(Query.schedule_resultset_size != None).order_by(Query.schedule_resultset_size.desc())
+        # Multiple queries with the same text may request multiple result sets
+        # be kept. We start with the one that keeps the most, and delete both
+        # the unneeded bridge rows and result sets.
+        first_query = queries.first()
+        if first_query is not None and first_query.schedule_resultset_size:
+            resultsets = QueryResultSet.query.filter(QueryResultSet.query_rel == first_query).order_by(QueryResultSet.result_id)
+            resultset_count = resultsets.count()
+            if resultset_count > first_query.schedule_resultset_size:
+                n_to_delete = resultset_count - first_query.schedule_resultset_size
+                r_ids = [r.result_id for r in resultsets][:n_to_delete]
+                delete_count = QueryResultSet.query.filter(QueryResultSet.result_id.in_(r_ids)).delete(synchronize_session=False)
+                QueryResult.query.filter(QueryResult.id.in_(r_ids)).delete(synchronize_session=False)
+        # Delete unneeded bridge rows for the remaining queries.
+        for q in queries[1:]:
```

Comment:
I'm not sure I follow what's happening in this loop. My understanding is that we look at the query with the largest allowable query results and delete its stale results and bridge rows like you said. That is what happens in the code above.

Since other queries may have pointed to the stale results that were deleted, the bridge rows should be deleted for those too? That's what I would have expected this loop to be doing. But it seems to be looking at whether queries that have more stale results, similar to the check that was done for the `first_query`?

I think in general there is some code in this loop that seems to be similar to what is done to the `first_query` and I'm wondering if factoring some of it out into a separate function might make it clearer or if it's possible include the operations on `first_query` in this loop?

Reply:
By this point, when the loop starts, there are no stale result sets left. (I'll add a comment pointing this out.)

So all that has to be deleted are bridge rows for queries that have requested fewer resultsets be kept than the first one.